server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java

```diff
@@ -289,14 +295,33 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws
      */
     protected void extractInnerHitBuilders(Map<String, InnerHitContextBuilder> innerHits) {}
 
+    public static QueryBuilder parseTopLevelQuery(XContentParser parser) throws IOException {
+        FilterXContentParser parserWrapper = new FilterXContentParserWrapper(parser) {
+            int nestedDepth;
+
+            @Override
+            public <T> T namedObject(Class<T> categoryClass, String name, Object context) throws IOException {
+                T namedObject = getXContentRegistry().parseNamedObject(categoryClass, name, this, context);
+                if (categoryClass.equals(QueryBuilder.class)) {
+                    nestedDepth++;
+                    if (nestedDepth > maxNestedDepth) {
+                        throw new IllegalArgumentException(
+                            "The nested depth of the query exceeds the maximum nested depth for bool queries set in ["
+                                + INDICES_MAX_NESTED_DEPTH_SETTING.getKey()
+                                + "]"
+                        );
+                    }
+                }
+                return namedObject;
+            }
+        };
+        return parseInnerQueryBuilder(parserWrapper);
+    }
+
     /**
      * Parses a query excluding the query element that wraps it
      */
-    public static QueryBuilder parseInnerQueryBuilder(XContentParser parser) throws IOException {
-        return parseInnerQueryBuilder(parser, Integer.valueOf(0));
-    }
-
-    public static QueryBuilder parseInnerQueryBuilder(XContentParser parser, Integer nestedDepth) throws IOException {
+    protected static QueryBuilder parseInnerQueryBuilder(XContentParser parser) throws IOException {
```

Comment:
this is now protected, which makes sure that only query builders call it, to parse inner queries. Whenever a top level query needs to be parsed, the public variant needs to be called. The distinction becomes now more important than before as the top level parsing method wraps the parser and starts tracking the nested depth.